### DISCUSSION TITLE
inquiry about the shimming toolbox used for Philips MRI

### DISCUSSION THREAD

**Discussion by gxu32** [shimming-toolbox/shimming-toolbox#513] [Q&A] [answered]
Dear shimming tool developers,

Thank you for your kind sharing this useful shimming tool. We're attempting to use shimming toolbox to do B0 shimming according your tutorials. But I meet some problems. Could you please give me some advice?
 
We obtained the field image anatomical image from our 7T philips MRI (Achieva) and plan to do static B0 shimming, here are my questions:
(1) I wonder if I set the '--slices' to volume, this would be a static B0 shimming?
 
(2) I set the '--slices' to volume and selected 1 and 2 for the Scanner order, After running the B0 shimming, I get 8 text files which was named from coefs_coil0_ch0 to coefs_coil0_ch7, Does the sequence of these text files correspond to the sequence of x, y, z, z2, zx, zy, x2-y2, 2xy ?
 
(3) For the static B0 shimming, do we still need coil_profiles file to get the exact current value for each shimming coil? If yes, do you have any ideas about which sequence We should scan on the philips MRI to get the images with 2 different currents for each channel? So we can use the image to calculate the B0 coil profiles.

Thank you for your help!

**Comment by po09i** [comment]
I appreciate your inquiry! Hopefully I can address your questions.

1) Indeed, setting the "--slices" option to "volume" will perform static/volume-wise optimization.
2) The output coefficients for Philips are ordered as you suggested (in the order you enter them in the scanner). To output the coefficients in a single file, I would recommend using the "--output-file-format-scanner" option with "chronological-coil". The ordering will be the same as you suggested for this option as well.
3) By using Shimming Toolbox's built-in scanner order functionality, the theoretical spherical harmonics are used. The coil profiles of the different spherical harmonic will be slightly different experimentally than theoretically. Therefore, you could map them to get better accuracy. However, the theoretical ones do a good job for us at the moment and I would recommend to try it out that way first (i.e.: you do not need to map them). 

To change the coefficients, we have been using the "manual" shim settings in  "Contrast"—>"Shim"—> "manual". I believe the "shimtool" EFO would also allow to input these coefficients.

Note that we have started implementation for Philips scanners recently and we are actively improving the process for Philips but here are some limitations as of right now.
- The table position must be at 0 (isocenter).
- The field map must be acquired with known coefficients. We currently do that by setting the coefficients to 0s. Shimming Toolbox can then process the coefficients which are added to the coefficients previously used for the field map (0s in this example, but they could be something else if you were doing more than 1 round of shimming). In other words, Shimming Toolbox outputs the change in coefficients, not the coefficients themselves.
- The shimming Toolbox currently supports up to the 2nd order on Philips. I would love to implement the 3rd order if you have data about them (field maps along with the acquired manual shim value used). 

If you decide to map the spherical harmonics, you can acquire with any sequence capable of computing B0 maps (note that Shimming Toolbox needs phase images or phase difference images), making sure to acquire with different known shimming coefficients. You can again use the "manual" shim dropdown for that.

**Reply by gxu32** [reply]
Thank you so much for your prompt and detailed reply!

**Comment by gxu32** [comment]
Sorry to bother you again.
With your kind help, I can run the static B0 shimming successfully, the resulted figure shows that the shimmed B0 filed map is better. But the coefficients seem to be very different from the coefficients which are obtained from the MRCode tool (also a tool for B0 shimming). The figures below are the parameters in the GUI used for our Static B0 shimming:
![1](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/421f5a7f-7f0f-4c4f-bb6b-8a0ac11c644a)
![2](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/b535a752-db4e-4f67-b494-cb2918700cbd)

The figures below  are the coefficients, the upper figure is the coefficient from MRCode tool, the lower one is from shimming toolbox.
 ![Figure_1_B0 Field Distribution](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/14e1b693-3995-43d2-8906-566615159b2e)
coefficient from shimming toolbox:
![6](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/78586748-345f-4049-a0f1-e0b337185e39)

The figure below is the messages which appears when I run the B0 shimming in shimming toolbox, I don't know if it is these Warnings that make the coefficients looks different.
![3](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/74afab73-b883-42ec-afe3-4e92e1fcee15)

For the "ShimSetting tag is not available" warning, I find that in the json file of our filed map there isn't the tag of ShimSetting, I don't know if this matter? 
For the 'Unknown manufacturer PHILIPS, assuming RAS' warning, I find that we do have this tag in our json file as shown in the figure below:
![7](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/6a5c5134-04ac-47b1-a6e0-c0221f401cec)

And I looked up the python code,  in the 'shimmingtoolbox.coils.coil' file, I found these sentences:
![4](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/eca811c7-8769-4e0d-88bb-57944117a657)

because the varible of 'SHIM_CS' is like this:
![5](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/653ad6b7-07c1-4d61-b60d-6ab5d8c8f0ef)

The SHIM_CS for our philips MRI should be 'RPI' , But from the warning, we find the manufacturer (PHILIPS) has not been recognized correctly and the SHIM_CS was set to the default of 'RAS'. I don't know if this warning also makes our coefficients a little strange. 

Many Thanks for your help！

**Comment by po09i** [comment]
> Sorry to bother you again.

No worries at all!

Philips does not populate the relevant information to populate the "ShimSettings" tag which is why it is not included in the JSON sidecar. The reason for the warning is because it means that Shimming Toolbox can't make sure that the scanner's constraints will be respected. For scanners, this is usually not an issue as a good shim typically requires much less current than what the scanner can provide.

The Philips specific code was merged in the main branch (Jan 23rd 2024) and in the plugin (Jan 26th 2024). I believe the "warning" message you are getting about Philips not being implemented might be because you are using a version prior to when it was implemented. I suggest updating to the newest version. To do so, navigate to where you cloned the plugin, pull the changes and reinstall:

```
cd <fsleyes-plugin-shimming-toolbox>
git pull
make install
```

You can then retry the command you used.

Let me know if that fixes your issue.
Alex :)

**Reply by gxu32** [reply]
Hi Alex,
Thank you for your prompt reply, I am working on it.

**Reply by gxu32** [reply]
Dear experts of shimming toolbox,

With your kind help, We finally get our static B0 shimming result, but still have a little problem. The following figure is the histogram of B0 fieldmap before shimming:
![beforeshim_histogram_simmingtoolbox](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/57b9d4dc-2cb1-4cf2-b816-5bc684ad2829)

The following one is the histogram of B0 filedmap after shimming using shimming toolbox:
![aftershim_histogram_simmingtoolbox](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/ab5b201b-7e9d-4503-86bf-5d38b885a075)
the histogram of the B0 filedmap after shimming seems not symmetrically converge. And the mean is -19.3 Hz, standard deviation values is about 127.6 Hz.

the following figure is the result from an another shimming tool (MRCodetool):
![Figure_1_B0 Field Distribution](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/a2d74181-815e-4b54-ab3d-b676899c59f1)

Below is the Sagittal B0 field image after shimming using shimming toolbox:
![Shimmingtoolbox_after_shim](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/01596305-457f-4db7-8d2b-28c274ca1b75)

Below is the Sagittal B0 field image after shimming using MRCodetool:
![MRCodetool_after_shim](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/ca894cd8-bc32-47b7-89df-285eedf02ca1)

It seems the bottom of the brain on the image of shimming toolbox is less homogeneous. The mask used for both tools are the same, maybe the erosion extent is a little different. I have tried several parameter combinations in the shimming toolbox, but the standard deviation values for these combinations are still relatively big, which is around 128 Hz. Could you give me some advices about this problem? Many thanks for your kind help!

Looking forward to your reply.
best,

**Reply by po09i** [reply]
Hi @gxu32! 

I have a couple of follow ups questions and comments. 

- Are the histograms of the shimmed B0 field map taken after experimentally acquiring the data with the new shim values or are these the histograms of the simulated shimmed B0 field map? (For both Shimming Toolbox and for MrCodeTool)
- The B0 field map you acquired using Shimming Toolbox seems to have an artifact at the bottom of the brain. I believe the B0 field should vary "more" smoothly in this region (similar to the field map you show when using MrCodeTool). Spherical harmonics also vary smoothly in space which would not produce this pattern. Could that be an issue with the field mapping process? (i.e.: Are the maps unwrapped correctly?)
- What are the masks that were taken to i) compute the histograms and ii) define the region of interest for shimming?. If the mask includes the artifact at the bottom, it could affect the histogram's shape.

I'm also curious of what the field map that was used to shim looked like. If there are also similar artifacts (wraps), it could affect the shimmed results.

The `pseudo_inverse` algorithm typically also performs slightly better than `least_square` (it is an unconstrained optimization however and does not respect bounds, but for scanner spherical harmonics, this is rarely an issue), it could help achieve better results if the issue is not the histogram's mask including the region with artifacts. 

Let me know if the problem truly is with phase wrapping and/if pseudo_inverse solves your issue and if I can be of any further help!

Best,
Alex

**Comment by gxu32** [comment]
Thanks for your reply! the following are the replies for the questions:

(1) the histogram is calculated from the simulated B0 filedmaps. I'm going to test it in our machine right now.

(2) because our machine can directly generate the B0 fieldmaps, So I just used its resulting B0 fieldmaps. Really appreciate your hints, I will try using the original phase images to get the B0 fieldmaps by the default and fsl’s unwrapping algorithm.

(3) the mask is also from our previous shimming tool (MRCode tool), because I want to see if how large the difference is between the two shimming tools. I will calculate a new mask and try B0 shimming again!  

I have aslo tried the pseudo_inverse, the result of the simulated B0 filedmap seems to be the same as the least_square. According to your suggestions, I am going to try them one by one.

Tnank you for your consistent help!
best regards,

**Reply by po09i** [reply]
You are very welcome!

Let me know if you have updates on your results. I'm curious how it compares to mrCode Tool and what we can do to improve Shimming Toolbox in the future. 

Best,
Alex

**Comment by po09i** [comment, marked as answer]
The issue was resolved in another discussion (#519). The main problem was that the B0 field map used for shimming needed to be unwrapped.
